dask_cuda/utils.py

```diff
@@ -21,3 +22,25 @@ def get_device_total_memory(index=0):
     """
     with cuda.gpus[index]:
         return cuda.current_context().get_memory_info()[1]
+
+
+def convert_frames_to_numba(frames):
+    try:
+        import cupy
+    except ImportError:
+        return frames
+
+    return [cuda.as_cuda_array(f) if isinstance(f, cupy.ndarray) else f for f in frames]
```

Comment:
Why the cupy check here?  I would expect numba to rely on the `__cuda_array_interface__` here rather than explicit type checking.  Does this not work today?  If not, can we raise an issue upstream?

Reply:
cuDF dataframes have various frames, some of them are not Numba device arrays. Example:

```python
In [1]: import pandas as pd, cudf

In [2]: from distributed.protocol import serialize

In [3]: pdf = pd.DataFrame({'a': [0, 1, 2, 3],'b': [0.1, 0.2, None, 0.3]})

In [4]: df = cudf.from_pandas(pdf)

In [5]: serialize(df, serializers=["cuda"])[1]
Out[5]:
[<numba.cuda.cudadrv.devicearray.DeviceNDArray at 0x7f58673875c0>,
 b'\x80\x04N.',
 <numba.cuda.cudadrv.devicearray.DeviceNDArray at 0x7f580ab45860>,
 b'\x80\x04N.',
 <numba.cuda.cudadrv.devicearray.DeviceNDArray at 0x7f580ab60e80>,
 <numba.cuda.cudadrv.devicearray.DeviceNDArray at 0x7f580ab607f0>]
```

This is why I'm avoiding traps of frames being non-device objects.

Reply:
In that case rather than type check, maybe we can check for the presense of `__cuda_array_interface__`.  In general I think that we should avoid explicit type checks when possible.  For example, if someone wants to use Dask with PyTorch or TensorFlow or some other CUDA library in the future, we shouldn't have to modify this code for them.

Perhaps we also record which of the frames we converted as a list of booleans, and include that information in the header.

Reply:
Yes, that's a good idea, thanks for the suggestion. I'll fix that.